### ISSUE TITLE
Infinite Loops and Hangs on CanvasKit Web 

### ISSUE THREAD

**Issue by jezell** [flutter/flutter#145380]
### Steps to reproduce

Occasionally when running into out of memory problems you will get stuck in an infinite loop in the WASM code. 

The problem is that CanvasKit is being compiled with aborting malloc enabled. However, CanvasKit uses Mutexes like this:

https://github.com/google/skia/blob/main/modules/skparagraph/src/ParagraphCache.cpp

This mutex will lock until its destructor is called. The problem is that if abort() is called, stack unwinding won't happen and the mutex won't call its destructor and CanvasKit will be left in an invalid state where the next call will lock the process. Normally, abort() would terminate the process, but it doesn't in WASM. 

The solution is to turn off aborting malloc, and turn on wasm exceptions, so that destructors are properly called, exceptions can be caught, and stack unwinding can happen. 

CanvasKit needs to be compiled with -fwasm-exceptions AND -sABORTING_MALLOC=0 in flutter web engine to fix these problems.

### Expected results

Don't go into infinite loops and leave mutexes locked.

### Actual results

App will hang if entering code locked by a mutex in canvaskit and then an exception occurs (OOM is easiest way to simulate).

### Code sample

See: https://github.com/flutter/flutter/issues/145134

It can be hard to repro because of other issues that cause flutter to stop working when unhandled exceptions happen in places that aren't guarded, but in a local build I've fixed a lot of those problems, and this hang be reproduced 100% of the time.

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[Paste your output here]
```

</details>

**Comment by mraleph** [comment]
> However, CanvasKit uses Mutexes like this:

Did you want to link to a specific line here @jezell?

/cc @yjbanov @eyebrowsoffire

**Comment by jezell** [comment]
@mraleph the flags are defined here for canvaskit in the flutter build, and you can see it disables exception catching:

https://github.com/flutter/engine/blob/main/skia/modules/canvaskit/BUILD.gn#L162

Though I think the -fwasm-exceptions flag needs to be turned on for all dependencies at both compile and link time to work properly. A specific example in paragraphcache.cpp is:

https://github.com/google/skia/blob/af65631d3afc06d59ddf7dac158b28da322c97d0/modules/skparagraph/src/ParagraphCache.cpp#L288

```cpp
void ParagraphCache::reset() {
    SkAutoMutexExclusive lock(fParagraphMutex);
#ifdef PARAGRAPH_CACHE_STATS
    fTotalRequests = 0;
    fCacheMisses = 0;
    fHashMisses = 0;
#endif
    fLRUCacheMap.reset();
    fLastCachedValue = nullptr;
}
```

A little sample I put together to confirm this behavior between different flags:

```
#include <malloc.h>
#include <emscripten.h>

#include <iostream>
#include <stdexcept>
#include <vector>

using namespace std;


int times = 0;

class Test { 
    int a[10000];

    public:
Test() {
}    
~Test() {
    }
};


class Mutex { 
    public:
Mutex() {
}    
~Mutex() {
        times++;
    }
};
extern "C" EMSCRIPTEN_KEEPALIVE size_t catchtest() {
    
Mutex m = Mutex();
try { 
      
      for(int i = 0; i < 1*1024*1024*1024;i++) {
       Test* t=  new Test();
      }
    } catch(...) {
    }
      return times;
}


extern "C" EMSCRIPTEN_KEEPALIVE size_t failtest() {
   
Mutex m = Mutex();
      for(int i = 0; i < 1*1024*1024*1024;i++) {
        Test *t = new Test();
      }
     return times;
}

extern "C" EMSCRIPTEN_KEEPALIVE size_t count() {
    return times;
}


```

This will cause an OOM exception inside catchtest and failtest. The counter will increment in the destructor, so if you call Module._catchtest() and Module._failtest() in the browser console you can see the difference between the two:

```em++ test.cpp -fwasm-exceptions -gsource-map -sWASM -sABORTING_MALLOC=0```

```em++ test.cpp -fwasm-exceptions -gsource-map -sWASM -sABORTING_MALLOC=1```

With aborting malloc turned on, the code doesn't increment the counter or catch the exception, and you can't catch it outside either ```try { Module._failtest(); } catch { }``` will not stop the behavior. When you turn off aborting malloc, the sample will behave correctly.

**Comment by yjbanov** [comment]
Thanks for the deep dive! I thought Google C++ style forbid use of exceptions, but I'm not sure what style we're meant to use for any sort of "try-finally" logic. I guess we should change the flags.

**Comment by eyebrowsoffire** [comment]
Turning aborting malloc off and turning exception handling on for OOM situations has an actual performance cost, and I don't think we should enable that for release builds. Perhaps it makes more sense to do so for a debug build of Canvaskit/Skwasm, and somehow make that debug build available to users. In general, I don't think we should consider OOM a recoverable state, although it might make sense to make it more visible to developers as an error, at least while in debug mode.

**Comment by yjbanov** [comment]
Yeah, if it has performance cost then simply enabling exceptions is not worth it as this is already not recoverable. We should track down the memory leak and fix it.

**Comment by yjbanov** [comment]
I'm going to close this issue since it's not actionable for the purposes of fixing the memory leak. Let's open a new issue that has a repro of the leak.

**Comment by jezell** [comment]
@eyebrowsoffire the performance cost is fairly minimal, but I get that some people want the smallest possible and fastest bundle and aren't encountering this issue. I agree with the general concept of officially supporting different canvaskit builds depending on the release configuration. For example, with some memory tracing turned on in the debug build we saw some leaks within Flutter in the console, and it would be nice maybe if flutter automatically turned on some things in release mode. It would also be nice if this was just a conditional in the gn file so we could compile our own build with this turned on. 

That said, there's a larger question as to why Skia is even using a mutex here. The hang could also be fixed by not using a spinlock and making the mutex do nothing in the browser, as there is only a single thread to begin with and the hang would be entirely avoidable if the mutex did nothing instead of doing something.

The larger problem right now is it's impossible for to prevent OOM scenarios in the browser because you can't see whether loading another image is going to cause a problem based off of how much memory you are using. #145646 would help avoid these problems.

At least so far, none of the issues we've uncovered with the renderer that are caused by OOM scenarios and prevent it from working are things that shouldn't be fixed. They are almost all just places where there is a lack of defensive coding and a catch or a finally block is missing in case an error is thrown somewhere. It's easy to cause errors to happen in random places with OOM, but it's not like an exception couldn't be thrown in the same paths without OOM.

**Comment by jezell** [comment]
@eyebrowsoffire @yjbanov raising the issue with the Skia team to see what they think.

https://issues.skia.org/issues/332939934

**Comment by jezell** [comment]
@eyebrowsoffire also note that this isn't really a true OOM scenario since the browser has plenty of memory available. The problem is more that CanvasKit can only use 2GB of WASM heap, and it's really easy to hit that limit in the browser if you are building anything that processes images. Imagine an app like Figma or Photoshop built on top of Flutter where the resources used by the app itself aren't static. We don't like our app randomly crashing and would like to stop it. That said, randomly crashing would still be better than randomly hanging. At the moment, CanvasKit flutter doesn't choose to crash in this case, it chooses to lock the browser process in a hang and force the user to terminate their browser process to fix it.

**Comment by jezell** [comment]
@yjbanov @eyebrowsoffire  would also note that dart itself when compiled to WASM uses WebAssembly exceptions, and the cost of the exception is really in the throwing of the exception, which in this case only happens from the new operator. The very idea that CanvasKit can't flip a switch to toggle the compile to generate a WASM exception from the oom on new instead of an abort call, for performance reasons, is a bit silly. 

<img width="814" alt="Screenshot 2024-04-04 at 2 37 33 PM" src="https://github.com/flutter/flutter/assets/22264/27c0b4ff-03aa-448b-9e3d-0182a7c882b8">

**Comment by yjbanov** [comment]
Reopening this one. Thanks @mraleph for flying all the way across the ocean to explain this one to me. I got confused after the discussion veered into OOM and memory leak territory. The core issue seems to be that the mutex implementation in Emscripten is not right for single-threaded code. In the absence of a second thread, waiting on a locked non-reentrant mutex is guaranteed to hang the program, since nothing but the current waiting thread can unlock it.

`-fwasm-exceptions` and `-sABORTING_MALLOC=0` is _a_ solution, but I'm not yet fully convinced that it's _the_ solution. Another solution could be for Emscripten to employ a different strategy for mutexes when code is known to be executing in a single-threaded context. For example, it could throw a JS exception that can be caught, allow the thread to reenter the mutex, etc. I think we should get the Emscripten team's opinion before attempting Flutter-side solutions.

**Comment by eyebrowsoffire** [comment]
Emscripten actually does stub out mutexes so that they are no-ops in single-threaded mode (which CanvasKit uses) but Skia's `SkAutoMutexExclusive` actually implements its own mutex on top of semaphores, and I'm not sure why. See https://github.com/google/skia/blob/main/include/private/base/SkMutex.h and https://github.com/google/skia/blob/main/include/private/base/SkSemaphore.h

**Comment by yjbanov** [comment]
What I should say is that whatever the lowest level primitive is in Emscripten that leads to the infinite loop situation, that thing might need to be revised. If there's a solution at that level, then we won't need to chase all individual usages of mutexes, semaphores, `<atomic>`s, pthreads, etc across all C++ codebases that we compile for WebAssembly. The solution will apply evenly everywhere. But if there's no clean solution at Emscripten level, then yeah, maybe the next place to look is Skia.

**Comment by mraleph** [comment]
This is still an issue. (cc @kevmoo)

**Comment by kevmoo** [comment]
Closing this out per old chat with @yjbanov. Let's track specific issues with reproductions, like https://github.com/flutter/flutter/issues/168190

**Comment by mraleph** [comment]
No, the only way this getting closed is if somebody submits a PR fixing it. This issue is older, has actual relevant discussion and details and might have people subscribed to it. If anything you are free to close #168190 as a duplicate of this one.

**Comment by lologarithm** [comment]
I have an app we are using the wasm compiler for.

Before flutter 3.32.X I had a panel that worked just fine. 

Starting around 3.32.0 opening the panel before doing anything else in the app would cause the entire thing to just... hang. Completely non-interactable, no error logs, just... broken.  This only occured on Chrome (mac and linux). However, Firefox (both OS) and Safari (mac only) work fine.

Now in 3.32.5 I see an error log when this happens, it appears to be something from paragraph_layout eventually ending in a malloc failure. 

I assume this is related to this whole issue. If not I can open a separate issue. 

If this is part of this issue, here is the stack trace (not sure if this will help at all):
```
skwasm.wasm:0xdf92 Uncaught 
$malloc	@	skwasm.wasm:0xdf92
$func460	@	skwasm.wasm:0x166e2
$func1643	@	skwasm.wasm:0x45b3e
$func5642	@	skwasm.wasm:0x12f8eb
$func10242	@	skwasm.wasm:0x26c836
$func10239	@	skwasm.wasm:0x26c417
$func8341	@	skwasm.wasm:0x1f95b4
$func2315	@	skwasm.wasm:0x6a82e
$func8338	@	skwasm.wasm:0x1f7a97
$func4725	@	skwasm.wasm:0xf9d03
$func8335	@	skwasm.wasm:0x1f779f
$func8577	@	skwasm.wasm:0x20640d
$paragraph_layout	@	skwasm.wasm:0x24a0b2
$func6535	@	main.dart.wasm:0x17ce69
```

I am willing to debug or try out dev builds of flutter to help fix this if needed.

Edit: I isolated the trigger of to a section of code that while animating changes font sizes rapidly with fractional values. I am working around by just rounding the font size instead of leaving it with all its decimal places. This seems to stop the hangs/crashes.

**Comment by kevmoo** [comment]
Thanks for the details @lologarithm – we'll take a look

**Comment by kevmoo** [comment]
@lologarithm – are you able to try the latest master channel? I believe we've fixed this!

**Comment by lologarithm** [comment]
Flutter 3.33.0-1.0.pre.817 • channel main • https://github.com/flutter/flutter.git
Framework • revision 750b87f551 (67 minutes ago) • 2025-07-03 14:09:23 -0400
Engine • revision 750b87f551 (67 minutes ago) • 2025-07-03 14:09:23 -0400
Tools • Dart 3.9.0 (build 3.9.0-293.0.dev) • DevTools 2.48.0-dev.0

The UI hang seems to be much harder to trigger now. But it still happens, also as I have played around with it more I haven't been seeing the paragraph_layout stack anymore and its back the rendering just dying (I can see debug output indicating that the app is still running though). It is possible these were two separate issues with the same result (UI hang) and you did in fact fix the paragraph_layout issue I found.

EDIT:: In fact, after the UI hang, I can still click on interaction spots I know were there and I can see from debug logs that the UI is still responding (so interaction events are still being handled). Its just the rendering is stopped entirely at that one moment during the animation where the panel is sliding.

**Comment by kevmoo** [comment]
@eyebrowsoffire just landed another memory fix. Could you take another look @lologarithm ?

**Comment by lologarithm** [comment]
I will try to get to that tonight or tomorrow. Thanks for updating!

**Comment by lologarithm** [comment]
Sorry for the long delay, been a busy summer :)

It seems that this hang is gone now, at least, enough gone that I can't get it to happen again (it was always kind of sporadic).  I will be sure to let you know if I find it again.Thank you @kevmoo
